### PR TITLE
Improve and correct vtkThreshold usage

### LINKED ISSUE #3610: Unexpected threshold behavior
### Describe the bug, what's wrong, and what you expected.

I'm using simple structed grids of cells, and need to filter-out some "nodata" cells. To do this, I'm setting scalar values to the cell data, then using [threshold](https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSetFilters.threshold.html) with the nodata value with `invert=True`. However, I'm getting confusing and inconsistent results compared to ParaView.

### Steps to reproduce the bug.

```python
import numpy as np
import pyvista

x = np.arange(5, dtype=float)
y = np.arange(6, dtype=float)
z = np.arange(2, dtype=float)
xx, yy, zz = np.meshgrid(x, y, z)
mesh = pyvista.StructuredGrid(xx, yy, zz)
mesh.cell_data.set_scalars(np.repeat(range(5), 4))

# All data
mesh.plot(show_edges=True)
# output is normal

# Filtering out nodata (zero) values
mesh.threshold(0, invert=True).plot(show_edges=True)
# output does not look normal, only 0-value cells are shown
```

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Thu Nov 17 15:23:57 2022 New Zealand Daylight Time

                OS : Windows
            CPU(s) : 12
           Machine : AMD64
      Architecture : 64bit
               RAM : 31.7 GiB
       Environment : IPython
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : NVIDIA RTX A4000/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 472.39

  Python 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:50:36) [MSC
  v.1929 64 bit (AMD64)]

           pyvista : 0.37.0
               vtk : 9.1.0
             numpy : 1.22.3
           imageio : 2.22.0
            scooby : 0.7.0
             pooch : v1.6.0
        matplotlib : 3.6.2
             PyQt5 : 5.12.3
           IPython : 8.6.0
          colorcet : 3.0.1
             scipy : 1.8.0
              tqdm : 4.63.0
            meshio : 5.3.4
--------------------------------------------------------------------------------
```

### Screenshots

Normal looking whole grid:
image

Odd-looking threshold attempt with pyvista, showing only 0-values:
image

Expected result with ParaView theshold filter with upper/lower set to 0 and invert selected:
image

Apologies for any "user error", as I'm new to this package.

### PR DESCRIPTION
**Pull request by banesullivan**
This should resolve #3610 

cc @mwtoews, would you please review if you can?

`vtkThreshold` has seen a few recent API changes, so we have to set the threshold limits and thresholding functions differently across VTK 8.1, 9.0, and 9.1+

This should keep backward compatibility (at least for our tests/examples)... there were obviously edge cases where the results were incorrect, and this should fix that. While this is labeled a breaking change for other reasons (see below on the `invert` parameter) and the results of this filter are indeed changing for many cases, they are changing in a correcting, rather than breaking, fashion.

@pyvista/developers, I'd appreciate a sanity check and close review to double-check what I've done here.

This also notes how we have a CELL preference for thresholding filters as opposed to a POINT preference pretty much everywhere else (ref #3744)

### Breaking Change: `invert` is no longer supported for VTK 8.1.2

Due to the fact that the invert option was not exposed on the `vtkThreshold` algorithm until VTK 9.0 and the complexity supporting this across 8.1.2, 9.0.x, and 9.1+, I am dropping support for the `invert` parameter for VTK<9 in PyVista. Thus the `invert` option is only valid for VTK 9+. To me, this seems like a fair compromise when trying to maintain support for an algorithm that has had significant API changes across multiple versions.

### Related

* https://github.com/pyvista/pyvista/issues/2850
* https://github.com/pyvista/pyvista/issues/3610
* https://discourse.vtk.org/t/unnecessary-vtk-api-change/9929

### CODE REVIEW

**Comment by banesullivan** [comment]
Ugh. This is a pain to maintain across versions going back to 8.1.2. `SetInvert` did not exist in 8.1.2

**Comment by banesullivan** [comment]
Seeing an error in the docs build that seems unrelated to these changes:

```
/home/runner/work/pyvista/pyvista/pyvista/core/filters/data_set.py:docstring of pyvista.core.filters.data_set.DataSetFilters.cell_centers:18: WARNING: Inline literal start-string without end-string.
```

File: pyvista/core/filters/data_set.py [hunk 1]
```diff
@@ -956,32 +957,44 @@ def threshold(
...
             maximum cell scalar] will be used to intersect the threshold bound,
             rather than the set of discrete scalar values from the vertices.
 
-        preference : str, optional
+        preference : str, default: 'cell'
             When ``scalars`` is specified, this is the preferred array
             type to search for in the dataset.  Must be either
-            ``'point'`` or ``'cell'``.
+            ``'point'`` or ``'cell'``. Throughout PyVista, the preference
+            is typically ``'point'` but since the threshold filter is a
```
Comment (MatthewFlamm) [inline pyvista/core/filters/data_set.py:995]:
```suggestion
            is typically ``'point'`` but since the threshold filter is a
```

File: pyvista/core/filters/data_set.py [hunk 2]
```diff
@@ -956,32 +957,44 @@ def threshold(
...
         invert : bool, optional
-            If value is a single value, when invert is ``True`` cells
-            are kept when their values are below parameter ``"value"``.
-            When invert is ``False`` cells are kept when their value is
-            above the threshold ``"value"``.  Default is ``False``:
-            yielding above the threshold ``"value"``.
+            Invert the threshold results. That is, cells that would have been
+            in the output with this option off are excluded, while cells that
+            would have been excluded from the output are included.
+            WARNING: This option is not supported for VTK<9.
```
Comment (mwtoews) [inline pyvista/core/filters/data_set.py:984]: this can be re-stated as a reST warning directive, e.g. [done here](https://github.com/pyvista/pyvista/blob/3db44d72070ccd6e8fc5c7f708500424011bcc47/pyvista/core/filters/poly_data.py#L2775)

File: pyvista/plotting/widgets.py [hunk 3]
```diff
@@ -1301,6 +1303,21 @@ def add_mesh_threshold(
             to intersect the threshold bound, rather than the set of
             discrete scalar values from the vertices.
 
+        all_scalars : bool, optional
```
Comment (mwtoews) [inline pyvista/plotting/widgets.py:1306]:
```suggestion
        all_scalars : bool, default: False
```

**Review by mwtoews — APPROVED** [review]
Thanks, testing with this PR works like a charm. Nice to see the deprecation warnings resolved! Only a few minor comments to consider.

File: pyvista/core/filters/data_set.py (hunk 1 above)
Comment (banesullivan) [reply pyvista/core/filters/data_set.py:995]: Ahhh this is where the warning is coming from. The warning was pointing me to the wrong spot

File: pyvista/core/filters/data_set.py (hunk 2 above)
Comment (banesullivan) [reply pyvista/core/filters/data_set.py:984]: Ah thanks! Will do

File: pyvista/plotting/widgets.py (hunk 3 above)
Comment (banesullivan) [reply pyvista/plotting/widgets.py:1306]: What's the preferred way to do this? IMO, bool, optional makes sense (being that default False is implied). This is the same for empty strings. But what's more "proper"? I think we do both throughout the codebase

Comment (MatthewFlamm) [reply pyvista/plotting/widgets.py:1306]: We wrote in the contributing section to use `optional` when the default is `None` to align  with `typing.Optional`, i.e. it doesn't match the type given for three partner, and `default: value`, when there is a default value that should match the type.

Comment (banesullivan) [reply pyvista/plotting/widgets.py:1306]: Ah I missed that. I may need to follow up and correct this in other places

**Comment by banesullivan** [comment]
@MatthewFlamm and @mwtoews, thank you for the quick review! I've modified the docstrings accordingly

**Review by mwtoews — DISMISSED** [review]

File: tests/plotting/test_widgets.py [hunk 4]
```diff
@@ -150,10 +150,12 @@ def test_widget_slider(uniform):
...
-    p = pyvista.Plotter()
-    p.add_mesh_threshold(uniform, invert=True)
-    p.add_mesh(uniform.outline())
-    p.close()
+    if pyvista.vtk_version_info >= (9,):
+        # Invert not support for VTK8.1.2
+        p = pyvista.Plotter()
+        p.add_mesh_threshold(uniform, invert=True)
+        p.add_mesh(uniform.outline())
+        p.close()
```
Comment (tkoyama010) [inline tests/plotting/test_widgets.py:158]: I guess we can improve the coverage of the test.
```suggestion
        p.close()
    else:
        # Invert not support for VTK8.1.2
        with pytest.raises(ValueError):
            p = pyvista.Plotter()
            p.add_mesh_threshold(uniform, invert=True)
```

File: tests/plotting/test_widgets.py [hunk 5]
```diff
@@ -150,10 +150,12 @@ def test_widget_slider(uniform):
     p.add_slider_widget(callback=func, rng=[0, 10], style="modern", pass_widget=True)
     p.close()
 
-    p = pyvista.Plotter()
-    p.add_mesh_threshold(uniform, invert=True)
-    p.add_mesh(uniform.outline())
-    p.close()
+    if pyvista.vtk_version_info >= (9,):
+        # Invert not support for VTK8.1.2
```
Comment (tkoyama010) [inline tests/plotting/test_widgets.py:154]:
```suggestion
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
LGTM with small suggestions.

File: tests/plotting/test_widgets.py (hunk 4 above)
Comment (banesullivan) [reply tests/plotting/test_widgets.py:157]: I'm actually going to revert this. This is hard to test for as the valueerror is being raised in a callback in a different event loop.

File: tests/test_filters.py [hunk 6]
```diff
@@ -450,6 +460,90 @@ def test_threshold_percent(datasets):
...
+    assert thresh.n_cells == 12
+    assert thresh.get_data_range() == (0, 2)
+    # lower(2),invert
+    #   [[      3, 3, 3],
+    #    [3, 4, 4, 4, 4]]
+    thresh = mesh.threshold(2, invert=True, method='lower')
+    assert thresh.n_cells == 8
+    assert thresh.get_data_range() == (3, 4)
+
+    # between(0, 0)
```
Comment (MatthewFlamm) [inline tests/test_filters.py:532]: This test is super thorough and well documented making it really easy to read.  Nice job!

A suggestion is to include a between that has a range.

File: pyvista/core/filters/data_set.py [hunk 7]
```diff
@@ -956,43 +957,54 @@ def threshold(
...
+           to prefer CELL data here so that if thresholding by a named array
+           that exists for both the POINT and CELL data, this filter will
+           default to the CELL data array while performing the CELL-wise
+           operation.
+
         Parameters
         ----------
         value : float or sequence, optional
-            Single value or (min, max) to be used for the data threshold.  If
+            Single value or (min, max) to be used for the data threshold. If
```
Comment (MatthewFlamm) [inline pyvista/core/filters/data_set.py:971]: I'm not sure what happens if the user passes a sequence in order (max, min).  I don't think this is explicitly handled here, and I'm not sure what vtk does with that (I assume it returns no cells?) Should we throw a warning?

**Review by MatthewFlamm — APPROVED** [review]
LGTM.  Some small suggestions that are not important to address.

File: tests/test_filters.py (hunk 6 above)
Comment (banesullivan) [reply tests/test_filters.py:532]: Ah, good catch. Done

File: pyvista/core/filters/data_set.py (hunk 7 above)
Comment (banesullivan) [reply pyvista/core/filters/data_set.py:971]: Hm. This will result in no-output. We should likely raise an Error rather than a warning. ParaView doesn't allow the user to do this

Comment (banesullivan) [reply pyvista/core/filters/data_set.py:971]: Added a `ValueError` for this

File: pyvista/core/filters/data_set.py [hunk 8]
```diff
@@ -5482,20 +5485,58 @@ def separate_cells(self):
...
+    if isinstance(value, (np.ndarray, collections.abc.Sequence)):
+        if len(value) != 2:
+            raise ValueError(
+                f'Value range must be length one for a float value or two for min/max; not ({value}).'
+            )
+    elif isinstance(value, collections.abc.Iterable):
+        raise TypeError('Value must either be a single scalar or a sequence.')
+    if pyvista.vtk_version_info >= (9,):
+        alg.SetInvert(invert)
+    elif invert:
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:5508]:
```suggestion
    elif invert: # pragma: no cover
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]

File: tests/plotting/test_widgets.py (hunk 4 above)
Comment (tkoyama010) [reply tests/plotting/test_widgets.py:158]: OK. Let's skip this using https://github.com/pyvista/pyvista/pull/3750#discussion_r1058441168 .

**Comment by banesullivan** [comment]
Merging after CI passes

File: pyvista/core/filters/data_set.py [hunk 9]
```diff
@@ -5482,20 +5485,58 @@ def separate_cells(self):
...
+        raise TypeError('Value must either be a single scalar or a sequence.')
+    if pyvista.vtk_version_info >= (9,):
+        alg.SetInvert(invert)
+    elif invert:  # pragma: no cover
+        raise ValueError('PyVista no longer supports inverted thresholds for VTK<9.')
+    # Check range
+    if isinstance(value, (np.ndarray, collections.abc.Sequence)) and value[0] > value[1]:
+        raise ValueError(
+            'Value sequence is invalid, please use (min, max). The provided first value is greater than the second.'
+        )
```
Comment (banesullivan) [inline pyvista/core/filters/data_set.py:5514]: Ehh... this might be cleaner to place under the previous `if isinstance(value, (np.ndarray, collections.abc.Sequence))` block

**Review by tkoyama010 — APPROVED** [review]
LGTM

**Review by MatthewFlamm — APPROVED** [review]
LGTM
